lib/config/presets/internal/helpers.ts

```diff
@@ -30,4 +30,15 @@ export const presets: Record<string, Preset> = {
       },
     ],
   },
+  pinGitHubActionDigestsToSemver: {
+    description: 'Convert pinned GitHub Action digests to SemVer.',
+    packageRules: [
+      {
+        extends: ['helpers:pinGitHubActionDigests'],
+        extractVersion: '^(?<version>v\\d+\\.\\d+\\.\\d+)$',
```

Comment:
Inside

https://github.com/renovatebot/.github/blob/95451c7263e7d6ef42198dceb188de271f8bedf0/default.json#L48

Reply:
This rule breaks actions like `rui314/setup-mold@v1`. They can no longer be pinned as there is no `v1.0.0` available at all.

Proposal to change this to:

`      "extractVersion": "^(?<version>v\\d+(?:\\.\\d+){0,2})$",`

That way `actions/checkout@v4` resolves all the way to `v4.1.1` and becomes pinned to that. (no change)

And for `rui314/setup-mold@v1` we detect `v1` as a valid version and allow pinning on that.

Example Repo: 

https://github.com/kristof-mattei/renovate-mold-no-version

PR to pin digests: https://github.com/kristof-mattei/renovate-mold-no-version/pull/4

PR that still updates actions/checkout@v4 to v4.1.1:

https://github.com/kristof-mattei/renovate-mold-no-version/pull/5

Reply:
Copy/pasting the proposal from @kristof-mattei so the maintainers and @TWiStErRob can review it:

```suggestion
        extractVersion: '^(?<version>v\\d+(?:\\.\\d+){0,2})$',
```

Reply:
LGTM syntactically and it seems meaningful.

I might ask consider simplifying it with capturing groups (no `?:`) and simple optional sequences `(...)?` just to use less regex features, which should equate to more people understanding it easier. Iff this doesn't break things for extraction.

How does this relate to https://github.com/renovatebot/renovate/pull/23663#discussion_r1281723709 where an executive decision was made for not supporting 2 components?

Reply:
> I might ask consider simplifying it with capturing groups (no `?:`) and simple optional sequences `(...)?` just to use less regex features, which should equate to more people understanding it easier. Iff this doesn't break things for extraction.

Can you suggest the correct regex for your idea?

I'll let @rarkins comment on the "executive decision they made about only supporting 3 version actions".

Reply:
For 1-2-3 components:
```suggestion
        extractVersion: '^(?<version>v\\d+(\\.\\d+)?(\\.\\d+)?)$',
```
For 1-3 components:
```suggestion
        extractVersion: '^(?<version>v\\d+(\\.\\d+\\.\\d+)?)$',
```

Reply:
@HonkingGoose I'll add one thing as to why I believe it makes sense to adopt this PR, but maybe under a different name.

I agree that v1 is not a semver version.

However, the reason I proposed this is because I had issues with `rui314/setup-mold` which only has a `v1` tag.

It seems that applying this ruleset disables the ability of the github-actions updater to consider versions like `v1`.

I believe that, even though `v1` isn't semver compatible, it is still valuable for me to be bale to pin `rui314@setup-mold@v1` to `rui314@setup-mold@sha-1234 #v1` AND `actions/checkout@v4` to `actions/checkout@sha-5678 # v4.1.1`

For the latter nothing changes.

But for the former now I can track the changes with both the sha & tag.

Reply:
@rarkins can you review this review thread and decide what you want to do? 😉

Reply:
It doesn't seem right to refer to it as "semver" if it's supporting short tags too. Is this then simply a "pin to longest version" PR, or is it more complicated than that?

Also keep in mind that this will be an opt-in preset so it doesn't _break_ anything, it simply may not solve all scenarios

Reply:
> It doesn't seem right to refer to it as "semver" if it's supporting short tags too. Is this then simply a "pin to longest version" PR, or is it more complicated than that?
> 
> Also keep in mind that this will be an opt-in preset so it doesn't _break_ anything, it simply may not solve all scenarios

You're right. My proposal clashes with the name. 

I'll propose a separate `pinToMostDetailedVersion`.

Reply:
FYI I haven't gotten the above to work — i.e. I tried this in my `renovate.json`:
```diff
-  }
+  },
+  "packageRules": [
+    {
+      "description": "When pinning GitHub Action digests, use most detailed version",
+      "matchDepTypes": ["action"],
+      "extractVersion": "^(?<version>v\\d+(\\.\\d+)?(\\.\\d+)?)$",
+      "versioning": "regex:^v(?<major>\\d+)(\\.(?<minor>\\d+)\\.(?<patch>\\d+))?$"
+    }
+  ]
 }
```
...and no improvement after letting Renovate retry/rebase my PRs.

Maybe I must use the `extends: ['helpers:pinGitHubActionDigests'],` for this to work at all — would they conflict?

Reply:
Ah, I just realized I had picked the "wrong" version, changing it now (in my repo's renovate.json) with:
```diff
-      "extractVersion": "^(?<version>v\\d+(\\.\\d+)?(\\.\\d+)?)$",
+      "extractVersion": "^(?<version>v\\d+\\.\\d+\\.\\d+)$",
```